src/platform/linux/x11/util/randr.rs

```diff
@@ -2,20 +2,40 @@ use std::{env, slice};
 use std::str::FromStr;
 
 use super::*;
-use super::ffi::{
-    RROutput,
-    XRRCrtcInfo,
-    XRRMonitorInfo,
-    XRRScreenResources,
-};
+
+pub fn calc_dpi_factor(
+    (width_px, height_px): (u32, u32),
+    (width_mm, height_mm): (u64, u64),
+) -> f64 {
+    // Override DPI if `WINIT_HIDPI_FACTOR` variable is set.
+    if let Ok(dpi_factor_str) = env::var("WINIT_HIDPI_FACTOR") {
+        if let Ok(dpi_factor) = f64::from_str(&dpi_factor_str) {
+            if dpi_factor <= 0. {
+                panic!("Expected `WINIT_HIDPI_FACTOR` to be bigger than 0, got '{}'", dpi_factor);
+            }
+            return dpi_factor;
+        }
+    }
+
+    // See http://xpra.org/trac/ticket/728 for more information.
+    if width_mm == 0 || width_mm == 0 {
+        return 1.0;
+    }
+
+    let ppmm = (
+        (width_px as f64 * height_px as f64) / (width_mm as f64 * height_mm as f64)
+    ).sqrt();
+    // Quantize 1/12 step size
+    ((ppmm * (12.0 * 25.4 / 96.0)).round() / 12.0).max(1.0)
```

Comment:
I don't really have a very deep understanding of this topic, but wouldn't the `96.0` here need to by the actual x11 screen DPI to properly support DPI on x11?

As far as I understand it 96.0 is only the default screen DPI on x11 and it could change based on a user's configuration. I'm not sure if in this context it's a different value that's always constant, but thought I'd just ask to properly understand this topic.

Reply:
Actually it might be possible that this `96.0` here is just used to calculate the DPI factor you'd need to apply to a screen with 96 DPI to get the "proper" DPI scaling. So this is probably all working as intended.

Reply:
I recommend reading the relevant issue #169 along with this blog post: http://wok.oblomov.eu/tecnologia/mixed-dpi-x11/

After that, you'll have an understanding with the same depth as mine, or probably better, since I know for a fact that I glossed over some things. I might not get a chance to look into that constant today, since I still have a Windows bug to investigate and a burning desire to relax for the rest of the day.

Reply:
Ah, GitHub only *sometimes* wants to update the page automatically, so I didn't see that. I do encourage being skeptical of this math, since I don't know who added it or when; I just know that it makes my eyes not hurt when using winit apps on my 1440p monitor.

Reply:
Thanks a lot for the clarification and the helpful links to the documentation. Until now I've mainly just made guesses based on the math in alacritty and winit and took a short read through the freetype documentation.

Looks like this will give me some more insights, very much appreciated!